### PR TITLE
added Order, OrderItem, service and other

### PR DESCRIPTION
**Pull request by RCherniavskiy**

### CODE REVIEW

File: src/main/java/book_store/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    @GetMapping("/{orderId}/items")
+    public ResponseEntity<List<OrderItemDto>> getAllOrderItemsForOrder(@PathVariable Long orderId) {
+        List<OrderItemDto> orderItems = orderService.getAllOrderItems(orderId);
+        return ResponseEntity.ok(orderItems);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(summary = "Retrieve a specific OrderItem within an order", description = "Endpoint for retrieving a specific OrderItem within an order")
+    @GetMapping("/{orderId}/items/{itemId}")
+    public ResponseEntity<OrderItemDto> getOrderItemForOrder(@PathVariable Long orderId, @PathVariable Long itemId) {
```
Comment (MAMentorFX) [inline src/main/java/book_store/controller/OrderController.java]: You dont need to return ResponseEntity. You have @ResponseStatus

File: src/main/java/book_store/mapper/OrderMapper.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Mapping(source = "user.id", target = "userId")
+    @Mapping(source = "orderItems", target = "orderItems")
+    OrderDto toDto(Order orders);
+
+    default OrderItemDto mapOrderItem(OrderItem orderItem) {
+        OrderItemDto orderItemDto = new OrderItemDto();
+        orderItemDto.setId(orderItem.getId());
+        orderItemDto.setQuantity(orderItem.getQuantity());
+        orderItemDto.setBookId(orderItem.getBook().getId());
```
Comment (MAMentorFX) [inline src/main/java/book_store/mapper/OrderMapper.java]: Use MapStruct please

File: src/main/java/book_store/mapper/OrderMapper.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import org.mapstruct.Mapping;
+import java.util.List;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderMapper {
+    @Mapping(source = "user.id", target = "userId")
+    List<OrderDto> toDtoList(List<Order> orders);
+
+    @Mapping(source = "user.id", target = "userId")
+    @Mapping(source = "orderItems", target = "orderItems")
```
Comment (MAMentorFX) [inline src/main/java/book_store/mapper/OrderMapper.java]:
```suggestion
```

File: src/main/java/book_store/service/OrderService.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.http.ResponseEntity;
+import java.util.List;
+
+public interface OrderService {
+    OrderDto orderReturn(OrderRequest requests, User user);
+
+    List<OrderDto> findAll(User user, Pageable pageable);
+
+    ResponseEntity<OrderDto> updateOrderStatus(Long id, String status);
```
Comment (MAMentorFX) [inline src/main/java/book_store/service/OrderService.java]:
```suggestion
    OrderDto updateOrderStatus(Long id, String status);
```

File: src/main/java/book_store/service/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        BigDecimal total = BigDecimal.ZERO;
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUser(user)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found for user: " + user.getUsername()));
+        Set<OrderItem> orderItems = new HashSet<>();
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
+            OrderItem orderItem = new OrderItem();
+            orderItem.setBook(cartItem.getBook());
+            orderItem.setQuantity(cartItem.getQuantity());
+            orderItem.setOrder(order);
+            orderItem.setPrice(cartItem.getBook().getPrice());
```
Comment (MAMentorFX) [inline src/main/java/book_store/service/OrderServiceImpl.java:46]: mapstrcuct

File: src/main/java/book_store/service/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        order.setTotal(total);
+        order.setOrderDate(LocalDateTime.now());
+        order.setOrderItems(orderItems);
+        order.setUser(user);
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Transactional
```
Comment (MAMentorFX) [inline src/main/java/book_store/service/OrderServiceImpl.java]: Why Transactional here?

File: src/main/java/book_store/service/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public List<OrderDto> findAll(User user, Pageable pageable) {
+        Page<Order> ordersPage = orderRepository.findByUser(user, pageable);
+        List<Order> orders = ordersPage.getContent();
+        return orderMapper.toDtoList(orders);
+    }
+
+    @Transactional
+    @Override
+    public ResponseEntity<OrderDto> updateOrderStatus(Long id, String status) {
```
Comment (MAMentorFX) [inline src/main/java/book_store/service/OrderServiceImpl.java]:
```suggestion
    public ResponseEntity<OrderDto> updateOrderStatus(Long id, Status status) {
```
Spring is able to convert string from json to enum

File: src/main/java/book_store/service/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Transactional
+    @Override
+    public List<OrderItemDto> getAllOrderItems(Long orderId) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order not found with id: " + orderId));
+        Set<OrderItem> orderItems = order.getOrderItems();
+        return orderItems.stream()
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (MAMentorFX) [inline src/main/java/book_store/service/OrderServiceImpl.java]: mapstruct is able to convert lists

File: src/main/java/book_store/service/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Transactional
```
Comment (MAMentorFX) [inline src/main/java/book_store/service/OrderServiceImpl.java]:
```suggestion
```

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/book_store/service/OrderServiceImpl.java (hunk 5 above)
Comment (RCherniavskiy) [reply src/main/java/book_store/service/OrderServiceImpl.java:46]: I don't see any point in this for two fields

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**